File: src/domain/captions.ts
import { z } from "zod";

const time = z.number().finite().nonnegative();
const color = z
  .string()
  .regex(/^#(?:[\da-f]{3}|[\da-f]{4}|[\da-f]{6}|[\da-f]{8})$/i, "Use a hex color such as #ffffff");
const epsilon = 0.00001;

export const captionWordSchema = z
  .object({
    text: z.string().trim().min(1).max(512),
    start: time,
    end: time,
    emphasis: z.boolean().default(false),
  })
  .superRefine((word, ctx) => {
    if (word.end <= word.start)
      ctx.addIssue({ code: "custom", message: "Word end must be after its start" });
  });

export const captionSegmentSchema = z
  .object({
    id: z.string().min(1).max(128),
    start: time,
    end: time,
    words: z.array(captionWordSchema).min(1).max(100),
  })
  .superRefine((segment, ctx) => {
    if (segment.end <= segment.start)
      ctx.addIssue({ code: "custom", message: "Caption end must be after its start" });
    segment.words.forEach((word, index) => {
      if (word.start < segment.start - epsilon || word.end > segment.end + epsilon) {
        ctx.addIssue({
          code: "custom",
          message: "Word timestamps must stay inside the caption",
          path: ["words", index],
        });
      }
      if (index > 0 && word.start < segment.words[index - 1].start) {
        ctx.addIssue({
          code: "custom",
          message: "Words must be ordered by start time",
          path: ["words", index],
        });
      }
    });
  });

export const captionStyleSchema = z.object({
  preset: z.enum([
    "classic",
    "bold",
    "karaoke",
    "highlight",
    "minimal",
    "podcast",
    "impact",
    "dynamic",
    "arabic",
  ]),
  fontFamily: z.string().trim().min(1).max(128),
  fontSize: z.number().finite().min(10).max(200),
  fontWeight: z.union([z.literal(400), z.literal(700), z.literal(900)]),
  color,
  highlightColor: color,
  outlineColor: color,
  outlineWidth: z.number().finite().min(0).max(12),
  shadow: z.boolean(),
  background: z.union([color, z.literal("transparent")]),
  x: z.number().finite().min(0).max(100),
  y: z.number().finite().min(0).max(100),
  lineHeight: z.number().finite().min(0.8).max(2),
  maxWidth: z.number().finite().min(10).max(100),
  direction: z.enum(["auto", "ltr", "rtl"]),
  animation: z.enum(["none", "fade", "pop"]),
  highlighting: z.enum(["none", "word", "karaoke"]),
});

export const captionTrackSchema = z
  .object({
    enabled: z.boolean(),
    style: captionStyleSchema,
    segments: z.array(captionSegmentSchema).max(10000),
  })
  .superRefine((captions, ctx) => {
    const ids = new Set<string>();
    captions.segments.forEach((segment, index) => {
      if (ids.has(segment.id))
        ctx.addIssue({
          code: "custom",
          message: "Caption IDs must be unique",
          path: ["segments", index],
        });
      ids.add(segment.id);
      if (index > 0 && segment.start < captions.segments[index - 1].end - epsilon) {
        ctx.addIssue({
          code: "custom",
          message: "Captions must be ordered without overlaps",
          path: ["segments", index],
        });
      }
    });
  });

export const transcriptSchema = z
  .object({
    language: z.string().min(1).max(64),
    words: z.array(captionWordSchema).max(100000),
  })
  .superRefine((transcript, ctx) => {
    if (
      transcript.words.some(
        (word, index) => index > 0 && word.start < transcript.words[index - 1].start,
      )
    ) {
      ctx.addIssue({ code: "custom", message: "Transcript words must be ordered by start time" });
    }
  });

export type CaptionWord = z.infer<typeof captionWordSchema>;
export type CaptionSegment = z.infer<typeof captionSegmentSchema>;
export type CaptionStyle = z.infer<typeof captionStyleSchema>;
export type CaptionTrack = z.infer<typeof captionTrackSchema>;
export type Transcript = z.infer<typeof transcriptSchema>;

export const defaultCaptions: CaptionTrack = {
  enabled: true,
  style: {
    preset: "classic",
    fontFamily: "Arial",
    fontSize: 64,
    fontWeight: 700,
    color: "#ffffff",
    highlightColor: "#fbbf24",
    outlineColor: "#000000",
    outlineWidth: 2,
    shadow: true,
    background: "transparent",
    x: 50,
    y: 78,
    lineHeight: 1.2,
    maxWidth: 86,
    direction: "auto",
    animation: "none",
    highlighting: "none",
  },
  segments: [],
};

export const captionPresets: { id: CaptionStyle["preset"]; name: string; style: CaptionStyle }[] = [
  { id: "classic", name: "Classic", style: { ...defaultCaptions.style } },
  {
    id: "bold",
    name: "Bold",
    style: {
      ...defaultCaptions.style,
      preset: "bold",
      fontSize: 78,
      fontWeight: 900,
      outlineWidth: 4,
    },
  },
  {
    id: "karaoke",
    name: "Karaoke",
    style: {
      ...defaultCaptions.style,
      preset: "karaoke",
      highlighting: "karaoke",
      highlightColor: "#4ade80",
    },
  },
  {
    id: "highlight",
    name: "Word Highlight",
    style: {
      ...defaultCaptions.style,
      preset: "highlight",
      highlighting: "word",
      highlightColor: "#facc15",
      animation: "pop",
    },
  },
  {
    id: "minimal",
    name: "Minimal",
    style: {
      ...defaultCaptions.style,
      preset: "minimal",
      fontSize: 46,
      fontWeight: 400,
      outlineWidth: 0,
      shadow: false,
      background: "#00000099",
    },
  },
  {
    id: "podcast",
    name: "Podcast",
    style: {
      ...defaultCaptions.style,
      preset: "podcast",
      fontFamily: "Georgia",
      fontSize: 60,
      background: "#0f172acc",
      outlineWidth: 0,
      animation: "fade",
    },
  },
  {
    id: "impact",
    name: "Impact",
    style: {
      ...defaultCaptions.style,
      preset: "impact",
      fontFamily: "Impact",
      fontSize: 88,
      fontWeight: 900,
      color: "#fbbf24",
      outlineWidth: 5,
      animation: "pop",
    },
  },
  {
    id: "dynamic",
    name: "Dynamic",
    style: {
      ...defaultCaptions.style,
      preset: "dynamic",
      fontSize: 74,
      fontWeight: 900,
      highlighting: "word",
      highlightColor: "#38bdf8",
      animation: "pop",
    },
  },
  {
    id: "arabic",
    name: "Arabic Viral",
    style: {
      ...defaultCaptions.style,
      preset: "arabic",
      fontFamily: "Tahoma, Arial",
      fontSize: 70,
      direction: "rtl",
      lineHeight: 1.5,
      highlighting: "word",
      animation: "fade",
    },
  },
];

export interface SegmentationOptions {
  /** Absolute start of the selected clip on the master timeline. */
  timelineStart: number;
  /** Duration of the selected source range; provider words are relative to this range. */
  duration: number;
  maxWords?: number;
  maxCharacters?: number;
  maxDuration?: number;
  silenceGap?: number;
}

/** Preserve recognizer word timestamps while grouping at punctuation, pauses, and reading limits. */
export function segmentTranscript(
  transcript: Transcript,
  options: SegmentationOptions,
): CaptionSegment[] {
  const input = transcriptSchema.parse(transcript);
  const parsed = z
    .object({
      timelineStart: time,
      duration: z.number().finite().positive(),
      maxWords: z.number().int().min(1).max(100).default(6),
      maxCharacters: z.number().int().min(1).max(51200).default(42),
      maxDuration: z.number().finite().positive().default(3),
      silenceGap: z.number().finite().nonnegative().default(0.65),
    })
    .parse(options);
  const words = input.words.flatMap((word) => {
    const start = Math.min(parsed.duration, word.start);
    const end = Math.min(parsed.duration, word.end);
    return end > start
      ? [{ ...word, start: start + parsed.timelineStart, end: end + parsed.timelineStart }]
      : [];
  });
  const segments: CaptionSegment[] = [];
  let group: CaptionWord[] = [];
  const flush = () => {
    if (!group.length) return;
    segments.push({
      id: crypto.randomUUID(),
      start: group[0].start,
      end: Math.max(...group.map((word) => word.end)),
      words: group,
    });
    group = [];
  };
  for (const word of words) {
    const previous = group.at(-1);
    const groupEnd = group.length ? Math.max(...group.map((item) => item.end)) : 0;
    const boundary =
      previous &&
      (group.length >= parsed.maxWords ||
        group.map((item) => item.text).join(" ").length + word.text.length + 1 >
          parsed.maxCharacters ||
        word.end - group[0].start > parsed.maxDuration ||
        word.start - previous.end >= parsed.silenceGap ||
        /[.!?؟؛…]["'”’»)]*$/u.test(previous.text));
    // Overlapping provider words stay in the same segment so segment ranges never overlap.
    if (boundary && word.start >= groupEnd) flush();
    if (group.length === 100)
      throw new Error(
        "Too many overlapping words in one caption. Correct the transcript timestamps.",
      );
    group.push(word);
  }
  flush();
  return captionTrackSchema.parse({ ...defaultCaptions, segments }).segments;
}

export function activeCaption(captions: CaptionTrack, time: number): CaptionSegment | null {
  if (!captions.enabled) return null;
  return captions.segments.find((segment) => segment.start <= time && time < segment.end) ?? null;
}

/** First strong letter sets the paragraph direction; digits and punctuation do not. */
export function captionDirection(
  text: string,
  direction: CaptionStyle["direction"],
): "ltr" | "rtl" {
  if (direction !== "auto") return direction;
  for (const character of text) {
    if (!/\p{Letter}/u.test(character)) continue;
    return /[\p{Script=Arabic}\p{Script=Hebrew}]/u.test(character) ? "rtl" : "ltr";
  }
  return "ltr";
}

export function captionWordIsHighlighted(
  word: CaptionWord,
  time: number,
  mode: CaptionStyle["highlighting"],
): boolean {
  return (
    word.emphasis ||
    (mode === "word"
      ? word.start <= time && time < word.end
      : mode === "karaoke" && word.start <= time)
  );
}

export function retimeCaptionSegment(
  segment: CaptionSegment,
  start: number,
  end: number,
): CaptionSegment {
  const scale = (end - start) / (segment.end - segment.start);
  return captionSegmentSchema.parse({
    ...segment,
    start,
    end,
    words: segment.words.map((word) => ({
      ...word,
      start: start + (word.start - segment.start) * scale,
      end: start + (word.end - segment.start) * scale,
    })),
  });
}

/** Split before a word, retaining every timestamp. A gap stays a gap between the captions. */
export function splitCaptionSegment(segment: CaptionSegment, wordIndex: number): CaptionSegment[] {
  if (!Number.isInteger(wordIndex) || wordIndex <= 0 || wordIndex >= segment.words.length)
    throw new Error("Choose a word after the first word to split the caption.");
  const first = segment.words.slice(0, wordIndex);
  const second = segment.words.slice(wordIndex);
  const firstEnd = Math.max(...first.map((word) => word.end));
  if (firstEnd > second[0].start)
    throw new Error("Adjust overlapping word times before splitting here.");
  return [
    { ...segment, end: firstEnd, words: first },
    { id: crypto.randomUUID(), start: second[0].start, end: segment.end, words: second },
  ];
}

File: src-tauri/src/captions.rs
//! Word timing and presentation persisted as part of the editable timeline.
use std::collections::HashSet;

use serde::{Deserialize, Serialize};

use crate::{error::AppResult, timeline::invalid};

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct CaptionWord {
    pub text: String,
    pub start: f64,
    pub end: f64,
    #[serde(default)]
    pub emphasis: bool,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct CaptionSegment {
    pub id: String,
    pub start: f64,
    pub end: f64,
    pub words: Vec<CaptionWord>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct CaptionStyle {
    pub preset: String,
    pub font_family: String,
    pub font_size: f64,
    pub font_weight: u32,
    pub color: String,
    pub highlight_color: String,
    pub outline_color: String,
    pub outline_width: f64,
    pub shadow: bool,
    pub background: String,
    pub x: f64,
    pub y: f64,
    pub line_height: f64,
    pub max_width: f64,
    pub direction: String,
    pub animation: String,
    pub highlighting: String,
}

impl Default for CaptionStyle {
    fn default() -> Self {
        Self {
            preset: "classic".into(),
            font_family: "Arial".into(),
            font_size: 64.0,
            font_weight: 700,
            color: "#ffffff".into(),
            highlight_color: "#fbbf24".into(),
            outline_color: "#000000".into(),
            outline_width: 2.0,
            shadow: true,
            background: "transparent".into(),
            x: 50.0,
            y: 78.0,
            line_height: 1.2,
            max_width: 86.0,
            direction: "auto".into(),
            animation: "none".into(),
            highlighting: "none".into(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct CaptionTrack {
    pub enabled: bool,
    pub style: CaptionStyle,
    pub segments: Vec<CaptionSegment>,
}

impl Default for CaptionTrack {
    fn default() -> Self {
        Self {
            enabled: true,
            style: CaptionStyle::default(),
            segments: vec![],
        }
    }
}

fn color(value: &str) -> bool {
    value.starts_with('#')
        && [4, 5, 7, 9].contains(&value.len())
        && value[1..].bytes().all(|byte| byte.is_ascii_hexdigit())
}

impl CaptionTrack {
    pub fn validate(&self) -> AppResult<()> {
        let style = &self.style;
        if self.segments.len() > 10_000
            || ![
                "classic",
                "bold",
                "karaoke",
                "highlight",
                "minimal",
                "podcast",
                "impact",
                "dynamic",
                "arabic",
            ]
            .contains(&style.preset.as_str())
            || style.font_family.trim().is_empty()
            || style.font_family.chars().count() > 128
            || ![
                style.font_size,
                style.outline_width,
                style.x,
                style.y,
                style.line_height,
                style.max_width,
            ]
            .iter()
            .all(|n| n.is_finite())
            || !(10.0..=200.0).contains(&style.font_size)
            || ![400, 700, 900].contains(&style.font_weight)
            || !color(&style.color)
            || !color(&style.highlight_color)
            || !color(&style.outline_color)
            || !(style.background == "transparent" || color(&style.background))
            || !(0.0..=12.0).contains(&style.outline_width)
            || !(0.0..=100.0).contains(&style.x)
            || !(0.0..=100.0).contains(&style.y)
            || !(0.8..=2.0).contains(&style.line_height)
            || !(10.0..=100.0).contains(&style.max_width)
            || !["auto", "ltr", "rtl"].contains(&style.direction.as_str())
            || !["none", "fade", "pop"].contains(&style.animation.as_str())
            || !["none", "word", "karaoke"].contains(&style.highlighting.as_str())
        {
            return Err(invalid(
                "invalid caption style or too many caption segments",
            ));
        }
        let mut ids = HashSet::new();
        let mut previous_end = 0.0;
        for segment in &self.segments {
            if segment.id.is_empty()
                || segment.id.len() > 128
                || !ids.insert(&segment.id)
                || !segment.start.is_finite()
                || !segment.end.is_finite()
                || segment.start < 0.0
                || segment.start < previous_end - 0.00001
                || segment.end <= segment.start
                || segment.words.is_empty()
                || segment.words.len() > 100
            {
                return Err(invalid(
                    "invalid, overlapping, or unordered caption segment",
                ));
            }
            validate_words(&segment.words, segment.start, segment.end)?;
            previous_end = segment.end;
        }
        // Captions remain editable after media trims; they never extend the media duration.
        Ok(())
    }
}

pub fn validate_words(words: &[CaptionWord], start: f64, end: f64) -> AppResult<()> {
    let mut previous_start = 0.0;
    for word in words {
        if word.text.is_empty()
            || word.text.trim() != word.text
            || word.text.chars().count() > 512
            || !word.start.is_finite()
            || !word.end.is_finite()
            || word.start < 0.0
            || word.start < start - 0.00001
            || word.start < previous_start
            || word.end <= word.start
            || word.end > end + 0.00001
        {
            return Err(invalid("invalid caption word text or timestamps"));
        }
        previous_start = word.start;
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    pub fn sample() -> CaptionTrack {
        CaptionTrack {
            segments: vec![CaptionSegment {
                id: "caption-1".into(),
                start: 1.0,
                end: 2.0,
                words: vec![CaptionWord {
                    text: "مرحبا".into(),
                    start: 1.0,
                    end: 1.5,
                    emphasis: true,
                }],
            }],
            ..Default::default()
        }
    }

    #[test]
    fn arabic_words_style_and_emphasis_round_trip() {
        let mut track = sample();
        track.style.direction = "rtl".into();
        track.validate().unwrap();
        assert_eq!(
            serde_json::from_str::<CaptionTrack>(&serde_json::to_string(&track).unwrap()).unwrap(),
            track
        );
    }

    #[test]
    fn rejects_overlap_duplicate_ids_and_unordered_segments() {
        let original = sample();
        let mut track = original.clone();
        track.segments.push(track.segments[0].clone());
        assert!(track.validate().is_err());
        track.segments[1].id = "caption-2".into();
        assert!(track.validate().is_err());
        track.segments[1].start = 0.0;
        assert!(track.validate().is_err());
    }

    #[test]
    fn rejects_invalid_word_times_and_untrimmed_text() {
        for word in [
            CaptionWord {
                start: 0.0,
                ..sample().segments[0].words[0].clone()
            },
            CaptionWord {
                end: 3.0,
                ..sample().segments[0].words[0].clone()
            },
            CaptionWord {
                end: f64::NAN,
                ..sample().segments[0].words[0].clone()
            },
            CaptionWord {
                text: "   ".into(),
                ..sample().segments[0].words[0].clone()
            },
        ] {
            let mut track = sample();
            track.segments[0].words[0] = word;
            assert!(track.validate().is_err());
        }
    }

    #[test]
    fn rejects_invalid_presentation_values() {
        let mut track = sample();
        track.style.font_size = f64::INFINITY;
        assert!(track.validate().is_err());
        track.style = CaptionStyle::default();
        track.style.color = "url(file://bad)".into();
        assert!(track.validate().is_err());
        track.style = CaptionStyle::default();
        track.style.direction = "sideways".into();
        assert!(track.validate().is_err());
    }

    #[test]
    fn missing_word_emphasis_defaults_to_false() {
        let word: CaptionWord =
            serde_json::from_str(r#"{"text":"Hello","start":0,"end":1}"#).unwrap();
        assert!(!word.emphasis);
    }

    #[test]
    fn timestamp_rounding_uses_the_frontend_tolerance() {
        let mut track = sample();
        track.segments[0].words[0].start = 1.0 - 0.000001;
        track.segments[0].words[0].end = 2.0 + 0.000001;
        let mut second = track.segments[0].clone();
        second.id = "caption-2".into();
        second.start = 2.0 - 0.000001;
        second.end = 3.0;
        second.words[0].start = 2.0;
        second.words[0].end = 2.5;
        track.segments.push(second);
        track.validate().unwrap();
        track.segments[1].start = 1.99;
        assert!(track.validate().is_err());
    }

    #[test]
    fn word_start_order_is_strict_even_inside_the_segment_tolerance() {
        let mut track = sample();
        let mut second = track.segments[0].words[0].clone();
        second.start -= 0.000001;
        track.segments[0].words.push(second);
        assert!(track.validate().is_err());
    }
}
